File: android/app/src/main/java/com/openledger/android/ui/auth/ApiKeySetupViewModel.kt
package com.openledger.android.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openledger.android.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ApiKeySetupState(
    val keyName: String = "Android Phone",
    val isLoading: Boolean = false,
    val error: String? = null,
)

@HiltViewModel
class ApiKeySetupViewModel @Inject constructor(
    private val authRepo: AuthRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(ApiKeySetupState())
    val state: StateFlow<ApiKeySetupState> = _state.asStateFlow()

    fun onKeyNameChanged(name: String) {
        _state.value = _state.value.copy(keyName = name, error = null)
    }

    fun createKey(onSuccess: () -> Unit) {
        _state.value = _state.value.copy(isLoading = true, error = null)
        viewModelScope.launch {
            authRepo.createApiKey(_state.value.keyName)
                .onSuccess { onSuccess() }
                .onFailure { e ->
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = "Failed to create key: ${e.message}",
                    )
                }
        }
    }
}

File: ios/OpenLedger/OpenLedger/ViewModels/ApiKeySetupViewModel.swift
import Foundation

@Observable
final class ApiKeySetupViewModel {

    var keyName: String = "iPhone"
    var isLoading: Bool = false
    var error: String?

    func createKey(authRepo: AuthRepository, onSuccess: @escaping () -> Void) {
        guard !keyName.trimmingCharacters(in: .whitespaces).isEmpty else {
            error = "Please enter a device name."
            return
        }

        isLoading = true
        error = nil

        Task { @MainActor in
            do {
                try await authRepo.createApiKey(name: keyName)
                isLoading = false
                onSuccess()
            } catch {
                isLoading = false
                self.error = "Failed to create key: \(error.localizedDescription)"
            }
        }
    }
}
